## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [33]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [34]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [35]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [36]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [37]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [38]:
# For the category with the most entries, which subcategory had the most entries?
df.loc[df.category == "consumables"].subcategory.value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [39]:
# Which five clients had the most entries in the data?
top_5_client = df.client_id.value_counts().head(5)
top_5_client

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [40]:
# Store the client ids of those top 5 clients in a list.
top_5_client_list = top_5_client.index.to_list()
top_5_client_list

[33615, 66037, 46820, 38378, 24741]

In [41]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = df.client_id.value_counts().head(1).index.to_list()[0]
df.loc[df.client_id == top_client_id].qty.sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [42]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
clear_df = df.copy()
clear_df["line_subtotal"] = clear_df["unit_price"] * clear_df["qty"]
display_columns = ["unit_price", "qty", "line_subtotal"]
clear_df[display_columns].head(2)



unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [43]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
clear_df = clear_df.assign(total_weight=df.unit_weight * df.qty)
clear_df = clear_df.assign(shipping_price = [weight * 7 if weight > 50 else weight * 10 for weight in clear_df["total_weight"]])

display_columns = ["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]
clear_df[display_columns].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [44]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
clear_df["line_price"] = ((clear_df.line_subtotal + clear_df.shipping_price) * 109.25 /100).round(2)
display_columns = ["line_subtotal", "shipping_price", "line_price"]

clear_df[display_columns].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [45]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
clear_df["line_cost"] = clear_df.unit_cost * clear_df.qty + clear_df.shipping_price
clear_df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [46]:
# Create a column for the profit of each line using line cost and line price
clear_df["line_profit"] = clear_df.line_price - clear_df.line_cost
clear_df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [47]:
# Check your work using the totals above
orders = [2742071, 2173913, 6128929]

def print_method(id):
    total_price = clear_df.loc[clear_df["order_id"] == id]["line_price"].sum().round(2)
    print(f"Order {id} total: ${total_price}")

for order in orders:
    print_method(order)


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [48]:
clear_df.loc[clear_df.client_id == 33615]


first   last                                           job  \
5      Jessica  Reyes  Production designer, theatre/television/film   
12     Jessica  Reyes  Production designer, theatre/television/film   
724    Jessica  Reyes  Production designer, theatre/television/film   
840    Jessica  Reyes  Production designer, theatre/television/film   
1405   Jessica  Reyes  Production designer, theatre/television/film   
...        ...    ...                                           ...   
52678  Jessica  Reyes  Production designer, theatre/television/film   
53450  Jessica  Reyes  Production designer, theatre/television/film   
53678  Jessica  Reyes  Production designer, theatre/television/film   
53995  Jessica  Reyes  Production designer, theatre/television/film   
54301  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
5      5016915923  jessica.reyes@grimes.net      33615   1327730  2023-05-02   
12     5016915923  jessica.reyes@grimes.net      33615   6059361  2023-05-16   
724    5016915923  jessica.reyes@grimes.net      33615   7417275  2023-05-24   
840    5016915923  jessica.reyes@grimes.net      33615   1951089  2023-02-15   
1405   5016915923  jessica.reyes@grimes.net      33615   5615970  2023-04-19   
...           ...                       ...        ...       ...         ...   
52678  5016915923  jessica.reyes@grimes.net      33615   3699596  2023-01-10   
53450  5016915923  jessica.reyes@grimes.net      33615   9985826  2023-02-15   
53678  5016915923  jessica.reyes@grimes.net      33615   1630456  2023-02-24   
53995  5016915923  jessica.reyes@grimes.net      33615   2196755  2023-03-08   
54301  5016915923  jessica.reyes@grimes.net      33615   7417275  2023-05-24   

       order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
5              18        2023  ...    154.44        2.79   78            0   
12             20        2023  ...    198.38       27.51  568            2   
724            21        2023  ...    155.57        1.33   26            2   
840             7        2023  ...      7.18        3.11   36            3   
1405           16        2023  ...     25.80        2.26   17            4   
...           ...         ...  ...       ...         ...  ...          ...   
52678           2        2023  ...     30.02        1.10   83            5   
53450           7        2023  ...     29.36        1.30  122            3   
53678           8        2023  ...    166.37        1.45   31            0   
53995          10        2023  ...     12.82       11.84  663            1   
54301          21        2023  ...     13.96        3.32   18            0   

       line_subtotal  total_weight  shipping_price  line_price  line_cost  \
5           13481.52        217.62         1523.34    16392.81   13569.66   
12         130492.32      15625.68       109379.76   262060.25  222059.60   
724          5772.78         34.58          345.80     6684.55    4390.62   
840           327.24        111.96          783.72     1213.72    1042.20   
1405          591.94         38.42          384.20     1066.43     822.80   
...              ...           ...             ...         ...        ...   
52678        3437.86         91.30          639.10     4454.08    3130.76   
53450        5431.44        158.60         1110.20     7146.74    4692.12   
53678        5592.09         44.95          449.50     6600.44    5606.97   
53995       10959.39       7849.92        54949.44    72005.40   63449.10   
54301         280.08         59.76          418.32      763.00     669.60   

       line_profit  
5          2823.15  
12        40000.65  
724        2293.93  
840         171.52  
1405        243.63  
...            ...  
52678      1323.32  
53450      2454.62  
53678       993.47  
53995      8556.30  
54301        93.40  

[220 rows x 24 columns]

In [49]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

for client in top_5_client_list:
    spend = clear_df.loc[clear_df.client_id == client]["line_price"].sum().round(2)
    print(f"{client}: &{spend}")

33615: &8377308.52
66037: &10259514.79
46820: &9743794.36
38378: &12906550.87
24741: &82268892.02


In [50]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

def get_top_client_info():
    display_columns = ["qty", "shipping_price", "line_price", "line_cost", "line_profit"]

    data = []
    for id in top_5_client_list:
        temp_df = clear_df.loc[clear_df.client_id == id]
        row = temp_df[display_columns].sum(axis=0)
        newRow = pd.concat([pd.Series({"client_id": id}), row])
        data.append(newRow)

    new_df = pd.DataFrame(data)
    new_df['client_id'] = new_df['client_id'].astype(int)
    new_df['qty'] = new_df['qty'].astype(int)
    
    return new_df

top_five_client_df = get_top_client_info()
top_five_client_df

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [51]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_column = ["shipping_price", "line_price", "line_cost", "line_profit"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    return amount / 1_000_000

# Apply the currency_format_millions function to only the money columns. 
top_five_client_df[money_column] = top_five_client_df[money_column].apply(lambda x: currency_format_millions(x))

# Rename the columns to reflect the change in the money format. 
top_five_client_df = top_five_client_df.rename(columns={"client_id": "Client ID",
                                   	"qty": "Units",
                                    "shipping_price": "Shipping (millions)",
                                    "line_price": "Total Revenue (millions)",
                                    "line_cost": "Total Cost (millions)",
                                    "line_profit": "Total Profit (millions)"})

top_five_client_df


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [52]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_df = top_five_client_df.sort_values(by="Total Profit (millions)", ascending=False).copy()
final_df

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995